### PR TITLE
Add unstamped 3D point variable

### PR DESCRIPTION
**Pull request by jakemclaughlin6**

### CODE REVIEW

File: fuse_core/include/fuse_core/uuid.h [hunk 1]
```diff
@@ -168,6 +168,17 @@ namespace uuid
...
+
+    /**
+   * @brief Generate a UUID from a namespace string, a uint64_t, and an additional id
+   *
+   * Every unique timestamp and id pair will generate a unique UUID
+   *
+   * @param[in] namespace_string A namespace or parent string used to generate non-overlapping UUIDs
+   * @param[in] user_id          A uint64_t user generated id
+   * @return                     A repeatable UUID specific to the provided namespace and timestamp
+   */
```
Comment (svwilliams) [inline fuse_core/include/fuse_core/uuid.h]:
```suggestion
    /**
   * @brief Generate a UUID from a namespace string and user provided id
   *
   * Every unique id will generate a unique UUID
   *
   * @param[in] namespace_string A namespace or parent string used to generate non-overlapping UUIDs
   * @param[in] user_id          A uint64_t user generated id
   * @return                     A repeatable UUID specific to the provided namespace and id
   */
```

File: fuse_core/src/uuid.cpp [hunk 2]
```diff
@@ -96,6 +96,13 @@ UUID generate(const std::string& namespace_string, const ros::Time& stamp, const
   return generate(namespace_string, buffer.data(), buffer.size());
 }
 
+UUID generate(const std::string& namespace_string, const uint64_t& user_id)
+{
+  ros::Time id_as_time;
+  id_as_time.fromNSec(user_id);
+  return generate(namespace_string, id_as_time);
+}
```
Comment (svwilliams) [inline fuse_core/src/uuid.cpp]: Converting the user id into a timestamp just to extract the raw bytes back out seems inefficient
```suggestion
UUID generate(const std::string& namespace_string, const uint64_t& user_id)
{
  return generate(namespace_string, reinterpret_cast<const unsigned char*>(&user_id), sizeof(user_id));
}
```

File: fuse_variables/include/fuse_variables/point_3d_landmark.h [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef FUSE_VARIABLES_POINT_3D_LANDMARK_H
```
Comment (svwilliams) [inline fuse_variables/include/fuse_variables/point_3d_landmark.h:37]: Feel free to add an acknowledgement for yourself here

File: fuse_variables/include/fuse_variables/point_3d_landmark.h [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+#include <ostream>
+
+namespace fuse_variables {
+
+/**
+ * @brief Variable representing a 3D point landmark that exists across time.
+ *
+ * This is commonly used to represent locations of visual feature locations. The
+ * UUID of this class is constant after construction and dependent on a use
```
Comment (svwilliams) [inline fuse_variables/include/fuse_variables/point_3d_landmark.h]:
```suggestion
 * UUID of this class is constant after construction and dependent on a user
```

File: fuse_variables/include/fuse_variables/point_3d_landmark.h [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    archive& boost::serialization::base_object<FixedSizeVariable<SIZE>>(*this);
+    archive & id_;
+  }
+};
+
+} // namespace fuse_variables
+
+BOOST_CLASS_EXPORT_KEY(fuse_variables::Point3DLandmark);
+
+#endif // FUSE_VARIABLES_POSITION_3D_H
```
Comment (svwilliams) [inline fuse_variables/include/fuse_variables/point_3d_landmark.h]:
```suggestion
#endif  // FUSE_VARIABLES_POINT_3D_LANDMARK_H
```

File: fuse_variables/src/point_3d_landmark.cpp [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+         << "  data:\n"
+         << "  - x: " << x() << "\n"
+         << "  - y: " << y() << "\n"
+         << "  - z: " << z() << "\n";
+}
+
+} // namespace fuse_variables
+
+BOOST_CLASS_EXPORT_IMPLEMENT(fuse_variables::Point3DLandmark);
+PLUGINLIB_EXPORT_CLASS(fuse_variables::Point3DLandmark, fuse_core::Variable);
```
Comment (svwilliams) [inline fuse_variables/src/point_3d_landmark.cpp]:
```suggestion
PLUGINLIB_EXPORT_CLASS(fuse_variables::Point3DLandmark, fuse_core::Variable);
```

File: fuse_variables/include/fuse_variables/point_3d_landmark.h [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+#include <fuse_core/variable.h>
+#include <fuse_variables/fixed_size_variable.h>
+
+#include <boost/serialization/access.hpp>
+#include <boost/serialization/base_object.hpp>
+#include <boost/serialization/export.hpp>
+
+#include <ostream>
+
+namespace fuse_variables {
```
Comment (svwilliams) [inline fuse_variables/include/fuse_variables/point_3d_landmark.h]:
```suggestion
namespace fuse_variables
{
```

File: fuse_variables/include/fuse_variables/point_3d_landmark.h [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+/**
+ * @brief Variable representing a 3D point landmark that exists across time.
+ *
+ * This is commonly used to represent locations of visual feature locations. The
+ * UUID of this class is constant after construction and dependent on a use
+ * input database id. As such, the database id cannot be altered after
+ * construction.
+ */
+class Point3DLandmark : public FixedSizeVariable<3> {
```
Comment (svwilliams) [inline fuse_variables/include/fuse_variables/point_3d_landmark.h]:
```suggestion
class Point3DLandmark : public FixedSizeVariable<3>
{
```

File: fuse_variables/include/fuse_variables/point_3d_landmark.h [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+   * @brief The Boost Serialize method that serializes all of the data members
+   * in to/out of the archive
+   *
+   * @param[in/out] archive - The archive object that holds the serialized class
+   * members
+   * @param[in] version - The version of the archive being read/written.
+   * Generally unused.
+   */
+  template <class Archive>
+  void serialize(Archive& archive, const unsigned int /* version */) {
```
Comment (svwilliams) [inline fuse_variables/include/fuse_variables/point_3d_landmark.h]:
```suggestion
  void serialize(Archive& archive, const unsigned int /* version */)
  {
```

File: fuse_variables/include/fuse_variables/point_3d_landmark.h [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+    archive& boost::serialization::base_object<FixedSizeVariable<SIZE>>(*this);
+    archive & id_;
+  }
+};
+
+} // namespace fuse_variables
```
Comment (svwilliams) [inline fuse_variables/include/fuse_variables/point_3d_landmark.h]:
```suggestion
}  // namespace fuse_variables
```

File: fuse_variables/src/point_3d_landmark.cpp [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+#include <fuse_core/uuid.h>
+#include <fuse_variables/fixed_size_variable.h>
+#include <fuse_variables/stamped.h>
+#include <pluginlib/class_list_macros.h>
+
+#include <boost/serialization/export.hpp>
+
+#include <string>
+
+namespace fuse_variables
```
Comment (svwilliams) [inline fuse_variables/src/point_3d_landmark.cpp]:
```suggestion
namespace fuse_variables
```

File: fuse_variables/src/point_3d_landmark.cpp [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+namespace fuse_variables 
+{
+
+Point3DLandmark::Point3DLandmark(const uint64_t& landmark_id)
+    : FixedSizeVariable(fuse_core::uuid::generate(detail::type(), landmark_id)), id_(landmark_id) 
+{
+}
+
+void Point3DLandmark::print(std::ostream& stream) const {
```
Comment (svwilliams) [inline fuse_variables/src/point_3d_landmark.cpp]:
```suggestion
void Point3DLandmark::print(std::ostream& stream) const
{
```

File: fuse_variables/src/point_3d_landmark.cpp [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+#include <boost/serialization/export.hpp>
+
+#include <string>
+
+namespace fuse_variables 
+{
+
+Point3DLandmark::Point3DLandmark(const uint64_t& landmark_id)
+    : FixedSizeVariable(fuse_core::uuid::generate(detail::type(), landmark_id)), id_(landmark_id)
```
Comment (svwilliams) [inline fuse_variables/src/point_3d_landmark.cpp]:
```suggestion
    : FixedSizeVariable(fuse_core::uuid::generate(detail::type(), landmark_id)), id_(landmark_id)
```

File: fuse_variables/src/point_3d_landmark.cpp [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+         << "  uuid: " << uuid() << "\n"
+         << "  size: " << size() << "\n"
+         << "  landmark id: " << id() << "\n"
+         << "  data:\n"
+         << "  - x: " << x() << "\n"
+         << "  - y: " << y() << "\n"
+         << "  - z: " << z() << "\n";
+}
+
+} // namespace fuse_variables
```
Comment (svwilliams) [inline fuse_variables/src/point_3d_landmark.cpp]:
```suggestion
}  // namespace fuse_variables
```

File: fuse_variables/test/test_point_3d_landmark.cpp [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+  }
+
+    // Verify two positions with the different landmark ids  produce different uuids
+  {
+    Point3DLandmark variable1(0);
+    Point3DLandmark variable2(1);
+    EXPECT_NE(variable1.uuid(), variable2.uuid());
+  }
+
+}
```
Comment (svwilliams) [inline fuse_variables/test/test_point_3d_landmark.cpp]:
```suggestion
  }
}
```

File: fuse_variables/include/fuse_variables/point_3d_landmark.h [hunk 16]
```diff
@@ -0,0 +1,162 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2018, Locus Robotics
```
Comment (svwilliams) [inline fuse_variables/include/fuse_variables/point_3d_landmark.h]:
```suggestion
 *  Copyright (c) 2021, Locus Robotics
```

File: fuse_variables/src/point_3d_landmark.cpp [hunk 17]
```diff
@@ -0,0 +1,67 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2018, Locus Robotics
```
Comment (svwilliams) [inline fuse_variables/src/point_3d_landmark.cpp]:
```suggestion
 *  Copyright (c) 2021, Locus Robotics
```

File: fuse_variables/test/test_point_3d_landmark.cpp [hunk 18]
```diff
@@ -0,0 +1,155 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2018, Locus Robotics
```
Comment (svwilliams) [inline fuse_variables/test/test_point_3d_landmark.cpp]:
```suggestion
 *  Copyright (c) 2021, Locus Robotics
```

**Review by svwilliams — APPROVED** [review]
I fixed some minor linter errors and simplified the UUID construction.

Let me know if you want to add any license or author attribution to the Point3Landmark class.

File: fuse_variables/include/fuse_variables/point_3d_landmark.h (hunk 3 above)
Comment (jakemclaughlin6) [reply fuse_variables/include/fuse_variables/point_3d_landmark.h:37]: Just added this, not sure where in the doc its meant to be so I just added it at the top

Comment (svwilliams) [reply fuse_variables/include/fuse_variables/point_3d_landmark.h:37]: Yeah....I have no idea either. If you're happy, then I'm happy.

**Review by ayrton04 — APPROVED** [review]

**Pull request merged**